Source/WebCore/bindings/js/ScriptModuleLoader.cpp

```diff
@@ -299,15 +299,31 @@ static bool isWorkletOrServiceWorker(ScriptExecutionContext& context)
     return false;
 }
 
-JSC::JSInternalPromise* ScriptModuleLoader::importModule(JSC::JSGlobalObject* jsGlobalObject, JSC::JSModuleLoader*, JSC::JSString* moduleName, JSC::JSValue parameters, const JSC::SourceOrigin& sourceOrigin)
+JSC::JSInternalPromise* ScriptModuleLoader::importModule(JSC::JSGlobalObject* jsGlobalObject, JSC::JSModuleLoader*, JSC::JSString* moduleName, JSC::JSValue parametersValue, const JSC::SourceOrigin& sourceOrigin)
 {
     JSC::VM& vm = jsGlobalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
     auto& globalObject = *JSC::jsCast<JSDOMGlobalObject*>(jsGlobalObject);
 
     // https://html.spec.whatwg.org/multipage/webappapis.html#hostimportmoduledynamically(referencingscriptormodule,-specifier,-promisecapability)
     // If settings object's global object implements WorkletGlobalScope or ServiceWorkerGlobalScope, then:
-    if (isWorkletOrServiceWorker(m_context))
+    if (isWorkletOrServiceWorker(m_context)) {
+        scope.release();
         return rejectPromise(globalObject, TypeError, "Dynamic-import is not available in Worklets or ServiceWorkers"_s);
+    }
+
+    auto reject = [&](auto& scope) {
+        auto* promise = JSC::JSInternalPromise::create(vm, globalObject.internalPromiseStructure());
+        return promise->rejectWithCaughtException(&globalObject, scope);
+    };
+
+    auto getTypeFromAssertions = [&]() -> JSC::ScriptFetchParameters::Type {
+        auto assertions = JSC::retrieveAssertionsFromDynamicImportOptions(&globalObject, parametersValue, { vm.propertyNames->type.impl() });
+        RETURN_IF_EXCEPTION(scope, { });
+        if (!assertions.isEmpty())
+            return JSC::ScriptFetchParameters::parseType(assertions.get(vm.propertyNames->type.impl())).value_or(JSC::ScriptFetchParameters::Type::JavaScript);
```

Comment:
If `JSC::ScriptFetchParameters::parseType` returns `nullopt`, this shouldn't fallback to `JSC::ScriptFetchParameters::Type::JavaScript` but reject the promise.

Step 5 of https://html.spec.whatwg.org/#fetch-an-import()-module-script-graph says
> If the result of running the [module type allowed](https://html.spec.whatwg.org/#module-type-allowed) steps given _moduleType_ and _settings object_ is false, then run _onComplete_ given null, and return.

(until a few days ago, "run onComplete given null" was written as "asynchronously complete this algorithm with null".

And then step 6.2 of https://html.spec.whatwg.org/#hostimportmoduledynamically(referencingscriptormodule,-modulerequest,-promisecapability) rejects the promise if the result is null:
> If result is null, then set _promise_ to [a promise rejected with](https://webidl.spec.whatwg.org/#a-promise-rejected-with) a new [TypeError](https://tc39.es/ecma262/#sec-native-error-types-used-in-this-standard-typeerror).

I found this in https://github.com/web-platform-tests/wpt/pull/35949